### ISSUE TITLE
Amesos2:  MUMPS matrix sequential/distributed input format

### ISSUE THREAD

**Issue by gvilanovac** [trilinos/Trilinos#6996]
## Question
@trilinos/amesos2

According to the MUMPS Users' Guide the matrix of the linear system can be given either on the host processor (INCTL(18)=0) or distributed over all processors (INCTL(18)=1,2,3). However, after looking at Amesos2 code, it seems to me that this option is not available.

Are you planning to make this option available? Also, I'm wondering if MUMPS is faster when the matrix is centralized on the host processor. And if so, is the overhead of copying the matrix to the host less than the speed up of solving with all the information centralized? What about memory?

Thank you!

**Comment by jhux2** [comment]
@trilinos/amesos2

**Comment by srajama1** [comment]
@gvilanovac Do you want support of this in 1 MPI rank ? I am somewhat confused by the host terminology.

**Comment by gvilanovac** [comment]
@srajama1 Sorry about the terminology. I used the one in MUMPS Users' Guide, and it is a bit confusing. Let me see if I can rephrase it:

According to MUMPS documentation, you can give the matrix of your problem to MUMPS in two different ways:
1.  The matrix is given in one processor (what they call the _host processor_, typically rank 0). This is the default option (INCTL(18)=0) and the one that I think it is been used in Amesos2. It means that if your matrix is initially distributed in all processors, you will have to copy all the data to one processor and give it to MUMPS. Then MUMPS will solve the system in parallel. 
2. The matrix is given through all processors. There are different options (INCTL(18)=1, 2 or 3) depending on how you stored your matrix, being INCTL(18)=3 the recommended one (pages 26-28 of the [manual](mumps.enseeiht.fr/doc/userguide_5.2.1.pdf)). Using this option, if the matrix is already distributed in all processors, you don't need to copy the data to one processor as in the previous case. Each processor gives its corresponding information of the matrix to MUMPS. Then MUMPS will solve the system in parallel. 

So my questions are:
- Am I right that Amesos2 uses option 1 (INCTL(18)=0)?
- If so, are you planning to implement option 2 (INCTL(18)=3)?
- Do you know if option 1 is faster than option 2?
- If so, is the overhead of copying the matrix to one processor less than the speed up of solving with all the information in one processor as in option 1?
- Have you ever experienced memory-related issues due to copying the matrix to one processor?

Thanks!

**Comment by srajama1** [comment]
@gvilanovac  I don't think the developer who contributed MUMPS needed distributed version on MUMPS. As a result, I believe this was never implemented.

We could implement other options, if needed. We also encourage contributions from users. 

The typical reason we do one processor is because that is our common use case, as a coarse solve for multigrid or for direct solve of small problems. There is a overhead in copying, but for such small problems that fit in one rank's memory parallel scalability may not be great any way. Yes, when problems don't fit in one rank, the distributed options are not needed. However, for such large problems, direct solvers might be expensive too. These are very special cases.

**Comment by gvilanovac** [comment]
@srajama1 Thank you for the information. It would be great to have this feature on trilinos. 

I'm planning to test Mumps (without the Amesos2 interface) to study the performance of the solver in both scenarios: what is already done in amesos2 vs using the distributed matrix. I can share the results here if you think this is interesting for you.

**Comment by srajama1** [comment]
Yes, that would be interesting. Thanks !

**Comment by gvilanovac** [comment]
@srajama1 I have performed the tests I told you. 

I have used the following two meshes:

mesh | nElem | nPts | nDOF | NNZ
-- | -- | -- | -- | --
mesh_a | 386048 | 193536 | 580608 | 63956992
mesh_b | 1542144 | 772096 | 2316288 | 256126976

I have analyzed the time (in seconds) that mumps took to solve the linear system in three scenarios: 
1. Sequential input of the matrix, as it is done in Amesos2; that is, copying the matrix to one processor and solving the system.
2. Parallel input of the matrix; that is, giving the matrix to mumps in distributed form.
3. Parallel input of the symmetric matrix. Same as above, but giving to mumps the upper triangular part of the matrix and setting the symmetric flag of mumps.
In all cases, I used metis for the reordering.

The results for mesh_a are:
#Procs | Sequential input | Parallel input | Parallel symmetric input | Seq/Par | Seq/Par sym
-- | -- | -- | -- | -- | --
56 | 15.72 | 11.77 | 8.76 | 1.34 | 1.79
42 | 16.04 | 11.89 | 9.03 | 1.35 | 1.78
28 | 18.57 | 14.29 | 9.91 | 1.30 | 1.87
14 | 19.37 | 15.96 | 11.14 | 1.21 | 1.74
7 | 26.47 | 23.70 | 16.11 | 1.12 | 1.64
4 | 33.43 | 31.28 | 20.75 | 1.07 | 1.61
2 | 55.53 | 54.97 | 34.83 | 1.01 | 1.59
1 | 87.12 | 86.93 | 52.37 | 1.00 | 1.66

The results for mesh_b are:
#Procs | Sequential input | Parallel input | Parallel symmetric input | Seq/Par | Seq/Par sym
-- | -- | -- | -- | -- | --
56 | 75.54 | 59.30 | 42.51 | 1.27 | 1.78
42 | 80.46 | 62.51 | 46.08 | 1.29 | 1.75
28 | 97.20 | 79.34 | 54.83 | 1.23 | 1.77
14 | 109.68 | 97.15 | 68.64 | 1.13 | 1.60
7 | 167.99 | 162.60 | 103.57 | 1.03 | 1.62
4 | 224.39 | 216.42 | 133.69 | 1.04 | 1.68
2 | 424.22 | 416.45 | 240.86 | 1.02 | 1.76
1 | 688.15 | 684.23 | 385.45 | 1.01 | 1.79

Please, note that I'm new to mumps and I'm sure that both the installation and the parameters of the library can be fine-tuned to optimize these computations.

I hope this is useful for you

**Comment by gvilanovac** [comment]
@srajama1 Just a short update on this topic.

The people from MUMPS have released a new version 5.3.1 in which they improved significantly the scalability of the library. This, combined with some parameter changes in calling MUMPS, produced great improvements:

#Procs | Sequential input | Parallel input | Seq/Par | Seq/Par sym
-- | -- | -- | -- | --
56 | 15.72 | 3.80 | 4.14 | 3.80
42 | 16.04 | 0.00 | #DIV/0! | #DIV/0!
28 | 18.57 | 4.60 | 4.04 | 4.60
14 | 19.37 | 5.33 | 3.63 | 5.33
7 | 26.47 | 11.30 | 2.34 | 11.30
4 | 33.43 | 19.80 | 1.69 | 19.80
2 | 55.53 | 27.60 | 2.01 | 27.60
1 | 87.12 | 22.40 | 3.89 | 22.40


#Procs | Sequential input | Parallel input | Parallel symmetric input | Seq/Par | Seq/Par sym
-- | -- | -- | -- | -- | --
56 | 75.54 | 12.10 | 7.34 | 6.24 | 10.29
42 | 80.46 | 11.80 | 7.10 | 6.82 | 11.33
28 | 97.20 | 17.10 | 8.89 | 5.68 | 10.93
14 | 109.68 | 25.40 | 11.80 | 4.32 | 9.29
7 | 167.99 | 36.80 | 19.00 | 4.57 | 8.84
4 | 224.39 | 46.40 | 20.20 | 4.84 | 11.11
2 | 424.22 | 93.70 | 34.30 | 4.53 | 12.37
1 | 688.15 | 196.00 | 23.20 | 3.51 | 29.66

**Comment by srajama1** [comment]
@gvilanovac These are interesting numbers. Thank you ! We will prioritize adding parallel inputs to MUMPS.
